core/src/avm2/amf.rs

```diff
@@ -404,39 +436,54 @@ pub fn deserialize_value<'gc>(
             );
             VectorObject::from_vector(storage, activation)?.into()
         }
-        AmfValue::VectorObject(vec, ty_name, is_fixed) => {
+        AmfValue::VectorObject(id, vec, ty_name, is_fixed) => {
             let name = AvmString::new_utf8(activation.context.gc_context, ty_name);
             let class = alias_to_class(activation, name)?;
-            let storage = VectorStorage::from_values(
-                vec.iter()
-                    .map(|v| {
-                        deserialize_value(activation, v).map(|value| {
-                            // There's no Vector.<void>: convert any
-                            // Undefined items in the Vector to Null.
-                            if matches!(value, Value::Undefined) {
-                                Value::Null
-                            } else {
-                                value
-                            }
-                        })
-                    })
-                    .collect::<Result<Vec<_>, _>>()?,
+
+            // Create an empty vector, as it has to exist in the map before reading children, in case they reference it
+            let empty_storage = VectorStorage::new(
+                0,
                 *is_fixed,
                 Some(class.inner_class_definition()),
+                activation,
             );
-            VectorObject::from_vector(storage, activation)?.into()
+            let obj = VectorObject::from_vector(empty_storage, activation)?;
+            object_map.insert(*id, obj);
+
+            let new_values = vec
+                .iter()
+                .map(|v| {
+                    deserialize_value_impl(activation, v, object_map).map(|value| {
+                        // There's no Vector.<void>: convert any
+                        // Undefined items in the Vector to Null.
+                        if matches!(value, Value::Undefined) {
+                            Value::Null
+                        } else {
+                            value
+                        }
+                    })
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+
+            // Swap in the actual values
+            obj.as_vector_storage_mut(activation.context.gc_context)
+                .unwrap()
```

Comment:
nit: I believe we generally prefer `expect`, with a meaningful message.

Reply:
Actually, I don't think it's worth blocking on this, I see no way for this to fail, `obj` is created locally, right above this line.

Reply:
If you are also okay with leaving this as is, @Dinnerbone, I'd say go for the merge.

Reply:
Done